lib/vnet/daemon/common.go

```diff
@@ -46,6 +56,8 @@ func (c *Config) CheckAndSetDefaults() error {
 		return trace.BadParameter("missing DNS address")
 	case c.HomePath == "":
 		return trace.BadParameter("missing home path")
+	case c.ClientCred == nil:
+		return trace.BadParameter("missing client credentials")
```

Comment:
`c.ClientCred` is a pointer only because I wanted to have _some_ form of safety in case a developer forgets to pass the creds.

This will of course catch only instances of someone forgetting to pass the whole struct. The Obj-C code that actually grabs egid and euid doesn't use pointers. If someone forgets to pass the creds from there, we'll end up with egid and euid both being set to `0`, which defeats the purpose of this whole thing. OTOH, I didn't want to error if both are equal to `0`, because then it'd be impossible to execute VNet as root. It's definitely not a common scenario, but it's something that should be perfectly valid.

Reply:
This wouldn't help with knowing on the Go side if creds received from Obj-C belong to the root user or simply weren't set.

Unless you meant using a struct like this on the Obj-C side as well? This could work (with some boilerplate, because I want to avoid mixing C structs into Obj-C).

Reply:
I don't think you need to return a "valid" from the FFI side, unless it's possible to actually not have that info - surely the code _in the package_ can be assumed to not misbehave.

Reply:
As far as I understand, `valid` needs to originate in the place that actually grabs the euid & egid. Imagine there's no `valid` field on the FFI side, we do some refactoring and we remove the code that actually does `outClientCred->euid = [[daemonService clientCred] euid]`. How would the Go side know if `euid` was not set or if `tsh vnet` was started as root?

Reply:
I think you're being a bit paranoid (being the same package and all), but being paranoid doesn't mean that they are not actually out to get you. 👁️